### PR TITLE
Fix issue with _checkpoint_without_reentrant

### LINKED ISSUE #76737: had_cuda_in_fwd is not defined in _checkpoint_without_reentrant if preserve_rng_state is False
### 🐛 Describe the bug

If you run torch.utils.checkpoint.checkpoint with both use_reentrant and preserve_rng_state set to False, you get the following error: `UnboundLocalError: local variable 'had_cuda_in_fwd' referenced before assignment`

Looking at the code, it is clear that the problem is in _checkpoint_without_reentrant.

I will add a minimal example:

```
import torch
from torch.utils import checkpoint

c = torch.nn.Linear(10,10).cuda()
x = torch.rand(10, 10, requires_grad=True).cuda()

checkpoint.checkpoint(c.forward, x, use_reentrant=False, preserve_rng_state=False)
```

It does _not_ fail without cuda.

### Versions

Collecting environment information...
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.9.7 (default, Sep 16 2021, 13:09:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.13.0-37-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.3.109
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 2080 Ti
GPU 1: NVIDIA GeForce RTX 2080 Ti

Nvidia driver version: 510.47.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.20.3
[pip3] torch==1.11.0
[pip3] torch-cluster==1.5.9
[pip3] torch-geometric==2.0.3
[pip3] torch-scatter==2.0.9
[pip3] torch-sparse==0.6.12
[pip3] torch-spline-conv==1.2.1
[pip3] torch-tb-profiler==0.3.1
[pip3] torchaudio==0.11.0
[pip3] torchvision==0.12.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py39h7f8727e_0  
[conda] mkl_fft                   1.3.1            py39hd3c417c_0  
[conda] mkl_random                1.2.2            py39h51133e4_0  
[conda] mypy-extensions           0.4.3                    pypi_0    pypi
[conda] numpy                     1.20.3                   pypi_0    pypi
[conda] pytorch                   1.11.0          py3.9_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch-cluster             1.5.9                    pypi_0    pypi
[conda] torch-geometric           2.0.3                    pypi_0    pypi
[conda] torch-scatter             2.0.9                    pypi_0    pypi
[conda] torch-sparse              0.6.12                   pypi_0    pypi
[conda] torch-spline-conv         1.2.1                    pypi_0    pypi
[conda] torch-tb-profiler         0.3.1                    pypi_0    pypi
[conda] torchaudio                0.11.0               py39_cu113    pytorch
[conda] torchvision               0.12.0               py39_cu113    pytorch

cc @ngimel @pbelevich

### PR DESCRIPTION
**Pull request by jamaliki**
Fixes  #76737
I also added a test case for this bug.

### CODE REVIEW

File: test/test_utils.py [hunk 1]
```diff
@@ -259,6 +259,19 @@ def run_fn(input):
 
             self.assertEqual(grad_with_checkpointing, grad_no_checkpointing)
 
+    @unittest.skipIf(not HAS_CUDA, 'No CUDA')
+    def test_checkpoint_not_preserve_rng_state_and_without_reentrant(self):
+        inp = torch.randn(20000, device='cuda').requires_grad_()
```
Comment (albanD) [inline test/test_utils.py:264]: Could you make that input of size 2 please? no need for it to be so big!

File: test/test_utils.py [hunk 2]
```diff
@@ -259,6 +259,19 @@ def run_fn(input):
...
+        layer = torch.nn.Dropout()
+
+        def run_fn(input):
+            return layer(input)
+
+        out = checkpoint(run_fn, inp, use_reentrant=False, preserve_rng_state=False)
+        out.sum().backward()
+        assert True
```
Comment (albanD) [inline test/test_utils.py:272]: No need for this.
In general, we shouldn't have plain asserts in the test suite (you want self.assertTrue()).
Here, you can simply leave a comment saying "Make sure that this runs without error".

**Review by albanD — COMMENTED** [review]
Looks good!
Only two minor comments on the tests.

File: test/test_utils.py (hunk 1 above)
Comment (jamaliki) [reply test/test_utils.py:264]: Fair enough :) will do

File: test/test_utils.py (hunk 2 above)
Comment (jamaliki) [reply test/test_utils.py:272]: Yes that makes sense

**Review by albanD — APPROVED** [review]
Perfect, thanks!

**Comment by albanD** [comment]
Thanks for contributing!

**Comment by jamaliki** [comment]
Thanks for the quick turn around, @albanD !

**Comment by nyngwang** [comment]
@albanD Thank you for making my world better.

**Pull request closed without merging**